runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/BatchFlinkExecutableStageContext.java

```diff
@@ -69,32 +75,132 @@ public StateRequestHandler getStateRequestHandler(
   }
 
   @Override
-  protected void finalize() throws Exception {
+  public void close() throws Exception {
     jobBundleFactory.close();
   }
 
   enum BatchFactory implements Factory {
-    INSTANCE;
-
-    @SuppressWarnings("Immutable") // observably immutable
-    private final LoadingCache<JobInfo, BatchFlinkExecutableStageContext> cachedContexts;
-
-    BatchFactory() {
-      cachedContexts =
-          CacheBuilder.newBuilder()
-              .weakValues()
-              .build(
-                  new CacheLoader<JobInfo, BatchFlinkExecutableStageContext>() {
-                    @Override
-                    public BatchFlinkExecutableStageContext load(JobInfo jobInfo) throws Exception {
-                      return create(jobInfo);
+    REFERENCE_COUNTING {
+      private static final int TTL_IN_SECONDS = 30;
+      int MAX_RETRY = 3;
+
+      @SuppressWarnings("Immutable") // observably immutable
```

Comment:
Is this suppression still necessary with the new constant-specific implementation? Why use a constant-specific here? Do we intend to create multiple implementations? Note that the enum only exists in the first place to make the singleton serialization-safe.